### ISSUE TITLE
Progress Notes

### ISSUE THREAD

**Issue by SamCoVT** [SamCoVT/FOCAL65_KIM1#1]
### Current Status:
Source now assembles.

In zero page: TXTBEG (Pointer to beginning of stored program), PBADR (Address of start of user program), VARBEG (Pointer to beginning of var list), VARST (Points to end of vars on 'E ALL' command), and VAREND (Pointer to end of var list) are all off (higher) by exactly $F5.  This is likely OK, as it is more likely to be wasting RAM by skipping a few available bytes before starting the variables and user program, rather than having overlapping user program and FOCAL code.  It looks like $36FF (the address in the new version) is beyond the end of both the old and new versions, so I will leave this alone for now.

In the code starting at $2000, there is a difference in the location of the "C FOCAL-65 (V3D) 26-AUG-77" string of 9 bytes (newly assembled code is 9 bytes later near the very end) so there must be some extra bytes/instructions somewhere in the middle.  Those will need to be located and eliminated if possible.  There is relatively good agreement up until $31A5 or thereabouts.  It looks like code may have been added to save the input device, shown here, that is not in the original binary (marked with a *** below):
```
03019              .WORD EVALM1  
03020    3193  A2 2D                  LDX #LINENO	;GET LINE OR GROUP TO 'DO' BACK
03021    3195  20 59 28               JSR POPB2               
03022    3198  20 3F 28               JSR POPA	;GET 'GETLN'FLAGS BACK
03023    319B  48                     PHA		;SAVE ON STACK FOR LATER 
03024    319C  20 C3 2C               JSR POPIV	;GET VALUE OF '&0' AND POINTER TO IT
03025    319F  20 BB 2C               JSR PUSHIV	;SAVE FOR LATER (VALUE IS IN FAC2)
03026    31A2  20 A1 2C               JSR PUTVAR	;NOW SET '&0' TO ARG VALUE (IN FAC1)
03027    31A5  A5 24                  LDA INSW	        ;*** SAVE WHERE INPUT IS COMING FROM ***
03028    31A7  20 31 28               JSR PUSHA         ;*** (PROGRAM OR INPUT DEVICE)       ***
03029    31AA  A9 00                  LDA #$00	        ;*** AND FORCE IT TO BE PROGRAM      ***
03030    31AC  85 24                  STA INSW          ;***
03031    31AE  68                     PLA		;GET STATUS FLAGS RETURNED BY 'GETLN'
03032    31AF  AA                     TAX		;SAVE IN X REGISTER
03033    31B0  20 1C 28               JSR PUSHJ	;NOW PERFORM THE 'DO' OF THE LINE OR GROUP
```

The old code also has $BC more bytes on the end after the copyright message.  I'm not sure what that data is, but it may be initial string/variable data.  We may need to implement it with .BYTE statements, just to get to binary compatibility.

**Comment by SamCoVT** [comment]
It's worth noting that those extra instructions are likely the "9 bytes" offset near the end... which is a bit odd because it looks like it is pushing data onto a stack and there would need to be extra code somewhere to pop it back off?

**Comment by SamCoVT** [comment]
Commenting out those lines makes most of the code line up properly.  There is a difference of 4 in several byte values and one section (where the I/O for the KIM was patched in" that is different, starting at TVOUT) that has a chunk that is different.  Otherwise, things are quite close now.

**Comment by HansOtten** [comment]
The extra code at the end is coming from the way I did the dump, I dumped just too much. The bytes are irrelevant and can be ignored. 
Same for zeropage, I just dumped the whole zeropage after a test run of Focal. That is possible with the tape routines in the KIM-1, but not possible to load as papertape.  It is not possible to overwrite the part where the KIM monitor keeps the status and the papertape load will fail.
Zeropage is scarcely initialized in an not run version  as you can see in the source.  Limit at least to the parts of zeropage to the end of the zeropage usage as shown in the source, so you can generate loadable papertape.
In my version I limit binary compatibility to the interpreter, and have the source regenerate the zeropage. 
Only strange part are those bytes list above the tables. No idea why they are there, not referenced anywhere.  It seems there was some room left for patching in manually the KEYIN routine, in the end it was not needed.

**Comment by SamCoVT** [comment]
Hi @HansOtten , thanks for the extra info.  I had figured that there was extra in the PTP file and that it was a "make sure you have enough" effort to get it on tape.  It's nice to have that confirmed.  I think that may also explain the issue I was having trying to load that particular PTP file into your KIM-1 simulator.

That difference of 4 in several bytes turned out to be the difference between zero page and immediate addressing (missing # in multiple places), so it was amazingly useful to have a binary to compare to - it would have taken me forever to find those by debugging, but I found it fun to locate them by binary comparison and fix them.  Thanks again for making your dump available.  I'm down to just a small number of differences now, and I think most of it is the patch for the KIM-1 I/O routines - thank you also for providing the assembly for that, as it will save me looking up all of the opcodes.

In an effort to make it easier for people who might want to port it to different hardware (myself included), I might look into adding a routine on the end to initialize the zero page variables.  I think the user manual hinted that it might be useful to add that to the initialization routine.  Your notes on the I/O routines should be helpful for that as well.

I mainly only have time on weekends to work on it, so it will be slow going.  I'll make sure to add a link to your website near the top of the README.md, in case people end up at my github first for some reason, as you have collected some useful info there.  I was also given some notes (I think from some correspondence between Wayne Wall and David Hassler) about adding functions, including how to generate the hash (without having FOCAL65 do it and using a breakpoint) and an example of how to add FCAL to call assembly routines from FOCAL65.  The example code appears to be missing a label, but I think I can see where it branches to, as it just skips over error handling.  If you don't already have those notes and would like a copy, let me know.

**Comment by HansOtten** [comment]
Look in the articles of KIM/6502 user notes, there is an article on adding a routine to fill zeropage instead of loading a separate file.
And yes, the notes are welcome!

**Comment by HansOtten** [comment]
Note the papertape you refer to is not the original, the binary on my website is. The papertape is 'slightly' patched says Dave, but what he did, it is not documented.  
So do a binary compare to my bin file. http://retro.hansotten.nl/uploads/focal/kimfocalv3d.zip has the dump from 1978.

**Comment by SamCoVT** [comment]
I'm hoping to have time to check that all out this weekend - thanks for pointing me to the correct file on your website to compare to.  I've sent you those notes via email, so let me know if you don't see them.

**Comment by SamCoVT** [comment]
I've made some progress and now have FOCAL working enough that the T command works with strings and simple math, but the initial copyright prints as garbage and trying to list the program or add a line also results in garbage.  Seeing as the copyright is technically line 0.0 of every program, I think it's related to the pointers in zero page.

I thought it looked like the binary zero page I was working with (Dave's version) had the zero page initialization patch installed, so I added that patch (which also moves line 0.0 and may have caused the above issue).  Now that I'm a bit beyond that point, I think it only had some of the variables modified for a 16K KIM-1 system, but not the zero page initialization.  I may need to go backwards and get to binary compatibility (including generating the zero page at address $0000 instead of as a separate initialization routine) to see if I can fix it.  I do like the convenience of the zero page initialization routine, though, so I would like to get that working.

It also looks like the binary I was working from (Dave's) had the patch to eliminate the echo suppression and speed up terminal I/O, but Hans' binary does not appear to have this patch and more closely matches the scanned listing, so I will leave that unpatched for now.  Hans' verison appears to work fine in his KIM-1 simulator, as I can load his binaries for zero page and the program and it works fine.

Good progress and I think things are close now.

**Comment by SamCoVT** [comment]
Got everything working.  This is currently a completely unpatched version.  It does not include the patch for faster I/O (which bypasses the echo defeat scheme) and does not have the patch for systems with 16K of RAM.  Hoping for some people with real hardware to verify that it works (the FOCALNEW.PTP file should be the easiest to load into real hardware).

**Comment by HansOtten** [comment]
Congrats!

**Comment by SamCoVT** [comment]
I had to add the KIM-1 I/O patch to get this working properly on physical KIM-1 hardware.  Interestingly, the patch seems to have a bug - when reading a character with echo turned on, it doesn't do a CLC to signal the character was read correctly.  I modified the patch to do the CLC when echo is turned on (it was already doing it when echo was turned off), and now the I/O is working both in Hans' KIM1SIM simulator and on real hardware.
